Replace boost::filesystem with std::filesystem

kiminuo opened the pull request:
## Introduction

C++17 has introduced the filesystem library. [cppreference.com](https://en.cppreference.com/w/cpp/filesystem) describes very well the origin of the library:

> The filesystem library was originally developed as boost.filesystem, was published as the technical specification ISO/IEC TS 18822:2015, and finally merged to ISO C++ as of C++17. The boost implementation is currently available on more compilers and platforms than the C++17 library.

This draft PR was created to gain feedback and examine what would be necessary to switch from `boost::filesystem` to `std::filesystem` in bitcoin codebase.

Bitcoin codebase contains `src/fs.h` which is a wrapper for the [currently used filesystem library](https://github.com/bitcoin/bitcoin/blob/6762a627ecb89ba8d4ed81a049a5d802e6dd75c2/src/fs.h#L18). The first impression is that one may just replace `namespace fs = boost::filesystem;` with `namespace fs = std::filesystem;` and all will be good. Unfortunately, there are some differences between the [Boost's filesystem library](https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html) and the [c++17 filesystem library](https://en.cppreference.com/w/cpp/filesystem). A nice summary by Davis Herring can be read [here](https://stackoverflow.com/questions/40899267/how-similar-are-boost-filesystem-and-the-c-standard-filesystem-library).

## Differences between filesystem implementations that affects Bitcoin code

1. `boost::filesystem::system_complete()` was renamed to `std::filesystem::absolute()` \[[source](https://stackoverflow.com/a/46271698)\]

2. [boost::filesystem::absolute()](https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html#absolute) has second `const path& base` parameter which was dropped in [std::filesystem::absolute](https://en.cppreference.com/w/cpp/filesystem/absolute) (see [reasoning](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0492r2.html#US78)).

     Each such instance like [this one](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/blockchain.cpp#L2284) can be fixed by changing:

     `fs::path path = fs::absolute(request.params[0].get_str(), GetDataDir());`

     to

    `fs::path path = fs::absolute(GetDataDir() / request.params[0].get_str());` where `/` is function call of `operator/` function.

     Note: Notably `path("foo") / "c:/bar"` yields `c:/bar`. However, that's how `boost::filesystem::absolute(p, base_dir)` behaves too.

3. [boost::filesystem::path::imbue()](https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html#path-imbue) is not present in `std::filesystem::path`.

4. [boost::filesystem::unique_path()](https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html#unique_path) is not present in `std::filesystem`.

5. [boost::filesystem::equivalent(path1, path2)](https://github.com/boostorg/filesystem/blob/812cb9bad3bcbb2161c7cd2c1132b51c4b29c4b9/src/operations.cpp#L1241) differs from [std::filesystem::equivalent(path1, path2)](https://github.com/llvm/llvm-project/blob/00b161b8971bc6d3cb55f13502288b8fe0dbaa42/libcxx/src/filesystem/operations.cpp#L917) as the later reports *not supported* error if one of the files does not exist whereas Boost returns `false`.
    * See also [cppreference](https://en.cppreference.com/w/cpp/filesystem/equivalent).

## Support for C++17 `<filesystem`>

* Clang 7+ (see [[Phoronix](https://www.phoronix.com/scan.php?page=news_item&px=Filesystem-Lands-LLVM-Libcpp)][[git commit](https://github.com/llvm-mirror/libcxx/commit/a0866c5fb5020d15c69deda94d32a7f982b88dc9)])
* GCC 8.1+ ([[GCC status page](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017)] [[git commit](https://github.com/gcc-mirror/gcc/commit/3b90ed62fb848046ed7ddef07df7c806e7f3fadb)])
  * GCC 8.x requires linking with `-lstdc++fs`.

## Issues & PRs that may be linked with this PR

https://github.com/bitcoin/bitcoin/issues/13103, https://github.com/bitcoin/bitcoin/pull/13787, https://github.com/google/leveldb/pull/760, https://github.com/bitcoin/bitcoin/pull/6093

## TODO

* [x] Remove Boost's filesystem from https://github.com/bitcoin/bitcoin/blob/master/configure.ac (help is welcome)
  * Relevant PR: https://github.com/bitcoin/bitcoin/pull/7302 & https://github.com/bitcoin/bitcoin/commit/76ac35f36d87078da62f95b4a1167ec296e37363
  * Done in #19667.
* [ ] [Am I missing `-lstdc++fs`](https://www.bfilipek.com/2019/05/boost-to-stdfs.html)?
* [ ] What about https://github.com/bitcoin/bitcoin/pull/19933?
* [ ] Address this change https://github.com/bitcoin/bitcoin/pull/19960
* [ ] `unicodepath` test in  `src/test/dbwrapper_tests.cpp` seems to require some UTF-8 locale.
  * See also https://github.com/bitcoin/bitcoin/blob/ae9ee5bdb18c5babbfa2875ab694df118d6a1c1b/src/util/system.cpp#L1249
  * https://github.com/microsoft/STL/blob/2aa944beba9af1d72537fd5e35d402c8b830a5dc/stl/src/filesystem.cpp#L266 - C++17 in Microsoft C++ standard library
  * https://docs.microsoft.com/en-us/cpp/standard-library/filesystem?view=msvc-160
  * https://docs.microsoft.com/en-us/cpp/standard-library/file-system-navigation?view=msvc-160

## Inspiration

* https://github.com/ceptontech/filesystem
* https://github.com/ceztko/u8filesystem

kiminuo committed:
Modify fs.h to switch from boost::filesystem to std::filesystem

(cherry picked from commit b526d92bf6b3deecb8bb2a9f6fcc50476a28ea80)
(cherry picked from commit 29cd1a22327469dcd625d24c9a317be23144887f)

kiminuo committed:
Replace `fs::system_complete(path)` with `fs::absolute(path)`

Resources:

* https://stackoverflow.com/a/46271698
* https://www.bfilipek.com/2019/05/boost-to-stdfs.html#differences-found-systemcomplete

(cherry picked from commit 88bb4e1bdfc43a449ea9d62d9336de9e162babfd)
(cherry picked from commit fc0d5a4c1c7153167afbca028d543ea65701658c)

kiminuo committed:
Replace boost::filesystem::absolute(path p, path base) with `std::filesystem::absolute(base_dir / p)`

Resources:

* http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/p0492r2.html#US78
* https://stackoverflow.com/questions/40899267/how-similar-are-boost-filesystem-and-the-c-standard-filesystem-library

(cherry picked from commit 65970378ab8a4b3e9b906d455afc32422635dab2)
(cherry picked from commit 465645d3a3d87c19d9c0f6b5b8c7795fd99ece6e)

# Conflicts:
#	src/rpc/blockchain.cpp
#	src/wallet/walletutil.cpp

kiminuo committed:
Remove `fs::path::imbue()` call as it no longer exist in `std::filesystem`

(cherry picked from commit eb89f20e1a45f1c58ec7396b376b65c7ae50bf28)
(cherry picked from commit b24d14961cd93a3229e4983ee31f56277b612ed5)

kiminuo committed:
Replace `fs::unique_path()` with `fsbridge::unique_path()` for now to re-implement it later

Resources:

* https://stackoverflow.com/questions/43316527/what-is-the-c17-equivalent-to-boostfilesystemunique-path

(cherry picked from commit fdeb20204a23b71a268c88b9d52d8678ea5b140d)
(cherry picked from commit 1a09157f3ca66f2931a55cf94a25cfa9944079a4)

kiminuo committed:
Remove `BOOST_FILESYSTEM_C_STR` as we want to use the overload with `fs::path`

See (1) to understand what the macro `BOOST_FILESYSTEM_C_STR` does.

Resources:

(1) https://www.boost.org/doc/libs/1_70_0/boost/filesystem/fstream.hpp
(2) https://stackoverflow.com/questions/821873/how-to-open-an-stdfstream-ofstream-or-ifstream-with-a-unicode-filename/#54842261

(cherry picked from commit 7543252fe53f992f93d6af802a8f5f30d40cce15)
(cherry picked from commit a02bba11c22108ecc4a88dbc106c2b7693b4838a)

kiminuo committed:
Remove workaround for opening UTF-8 paths on Windows

(cherry picked from commit 6df4d92034b4302870162e108f48cefd4e21bd1a)
(cherry picked from commit f71054b2a407a089e1915b992a536c8731356922)

kiminuo committed:
Remove `boost/filesystem/fstream.hpp` from `EXPECTED_BOOST_INCLUDES` in `lint-includes.sh`

(cherry picked from commit 3358071a54c9d803f91b4daca00cf963b4390863)
(cherry picked from commit 8d1464d4513bf705a7ae904c96a768c9b08ce027)

kiminuo committed:
Replace `boost::system::error_code` with `std::error_code`

(cherry picked from commit 6d375ab584355a78ff2f2e347063b990b1573ee5)
(cherry picked from commit 593e7df400e0d71e1e355668ee747491fb60749b)

# Conflicts:
#	src/wallet/walletutil.cpp

practicalswift commented:
Concept ACK: this will be nice to have in the future when we switch to C++17. Thanks for experimenting!

MarcoFalke commented:
Would it make sense to split out the two `absolute()` commits, because they make sense even with C++11?

kiminuo committed:
Fix db test "getwalletenv_file"

(cherry picked from commit 2e19c4cbc29e9778c8d037961f156a5f1b7280d9)
(cherry picked from commit 1043687a555101c0b3550dde48bfd29c8e82c28f)

kiminuo committed:
Fix init_tests

(cherry picked from commit db44346266d71b716f136fe6d8febf819168ce27)
(cherry picked from commit 62e7fb353efbbc125b1a96ef78e0331149b144ca)

kiminuo commented:
> Would it make sense to split out the two `absolute()` commits, because they make sense even with C++11?

I think so. Yes.

laanwj commented:
Nice work.

kiminuo committed:
User u8path rather -- TODO: Verify!!

(cherry picked from commit 018574609b16ae45f2448f9437267ac68c1e70d7)
(cherry picked from commit 2d8823f1bdb530088e44c2cdca040ad146ce0752)

kiminuo committed:
db.cpp - change fs::copy_option::ovewrite_if_exists -> fs::copy_options::ovewraite_existing

(cherry picked from commit 5af0554ce526374af076e266ff46e29dff3bd5cc)

MarcoFalke commented on src/fs.cpp:
```diff
@@ -31,6 +32,14 @@ FILE *fopen(const fs::path& p, const char *mode)
+    boost::filesystem::path p = boost::filesystem::unique_path();
+    fs::path p2{p.string()};
+
+    return p2;
```
I think all that boost does it pick 64 bits of randomness? Something along the lines of 

```suggestion
    return g_insecure_rand_ctx_temp_path.rand256().ToString().substr(0, 16);
```

should be identical.

kiminuo replied:
Yes. I believe so too.

However, using this leads to a new circular dependency "fs -> random -> logging -> fs": https://travis-ci.com/github/kiminuo/bitcoin/jobs/360282950#L466 (https://github.com/kiminuo/bitcoin/commit/333ab6ad421f1b25bf66bd2557cd3d0e4b60e0f6)

`unique_path()` is used only twice but I don't know what is preferred solution here. Inline it? Move function somewhere else? Where?

_____
For other people who may read this: For [used](https://github.com/bitcoin/bitcoin/blob/master/doc/dependencies.md) Boost 1.70.0:

* Source code: https://github.com/boostorg/filesystem/blob/e49775e8bbc4d8dbea5fbf6cf7c29f4d7bbd0c67/src/unique_path.cpp#L131
* Docs: https://www.boost.org/doc/libs/1_70_0/libs/filesystem/doc/reference.html#unique_path

MarcoFalke replied:
I think there are two options:

* Inline
* or create a new header, e.g. `src/util/fs.h`

kiminuo committed:
Do not use Boost implementation of `fs::path::unique_path`.

Suggestion here: https://github.com/bitcoin/bitcoin/pull/19245#discussion_r451537286

elichai commented:
Nice! Please ping me when it's ready for review :)

kiminuo commented:
@theuni Hi, would you know how to remove this reference to Boost's filesystem: 

https://github.com/bitcoin/bitcoin/blob/2a784723f0c0e642353dc74ec6aef4d5f8345044/configure.ac#L1285

? 

It comes from #7302 (commit 76ac35f)

kiminuo committed:
Remove AX_BOOST_FILESYSTEM test.

kiminuo committed:
Remove suppressions for boost::filesystem from valgrind.supp file.

kiminuo committed:
depends: Remove Boost filesystem from config libraries.

kiminuo committed:
Do not require to install `libboost-filesystem-dev` in tests and in documentation.

# Conflicts:
#	build_msvc/README.md
#	build_msvc/vcpkg-packages.txt
#	ci/test/00_setup_env_native_asan.sh
#	ci/test/00_setup_env_native_valgrind.sh

kiminuo committed:
Modify fs::absolute call in ArgsManager::GetSettingsPath.

elichai commented:
@kiminuo Now that #20413 is merged, are you planning on rebasing this and getting it ready for review?

kiminuo commented:
> @kiminuo Now that #20413 is merged, are you planning on rebasing this and getting it ready for review?

I will get to it. However, @MarcoFalke mentioned on IRC briefly that `std::fs` is not available on some supported OSs even with C++17 enabled. I'm not sure I understood correctly. If true, it would delay this feature months or years.

kiminuo committed:
To split into multiple commits

kiminuo committed:
Fix tests.

kiminuo committed:
Fix util_tests.

kiminuo committed:
why?

kiminuo committed:
Test.

kiminuo committed:
Test.

kiminuo committed:
workaround: dbwrapper_tests

kiminuo committed:
Test with 20.04

kiminuo committed:
Test: Use Clang 10.

MarcoFalke commented:
Hi, catching up on the IRC log.

I just wanted to say that this is blocked on #20460 and there is nothing you can do about this in this pull. Fixing #20460 will need build system and gitian/guix, as well as CI changes. This would be too much to do here.

fanquake commented:
Thanks for your work here so far @kiminuo. I've addressed the build system requirements, and combined this changes into #20744 (still attributed to you). However due to runtime and compiler requirements it's unlikely we'll be able to start using them just yet. I'm going to close this PR in favour of #20744 for now.
